cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/metrics/MetricsHandlerTestRun.java

```diff
@@ -670,6 +738,221 @@ public void testSearchMetrics() throws Exception {
+    ));
+    batchTest(taggedQuery, expected);
+    taggedQuery = new JsonObject();
+    queries = new JsonArray();
+    queries.add(q1);
+    taggedQuery.add("testBatch1", queries);
+    queries = new JsonArray();
+    queries.add(q2);
+    taggedQuery.add("testBatch2", queries);
+
+    expected = ImmutableMap.of("testBatch1", ImmutableList.<ImmutableList<QueryResult>>of(
+                                 ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(),
+                                                                               "system.reads", 1, 4))),
+                               "testBatch2", ImmutableList.<ImmutableList<QueryResult>>of(
+        ImmutableList.<QueryResult>of(new QueryResult(ImmutableMap.<String, String>of(), "system.reads", 1, 2),
+                                      new QueryResult(ImmutableMap.<String, String>of(), "system.writes", 1, 2)))
+    );
+    batchTest(taggedQuery, expected);
+
+    // test invalid request - query without any query Params and body content
+    HttpResponse response = doPost("/v3/metrics/query", null);
+    Assert.assertEquals(400, response.getStatusLine().getStatusCode());
+  }
+
+  JsonObject getSingleQueryJson(Map<String, String> tags, List<String> metrics,
+                               List<String> groupBy, Map<String, String> timeRange) {
+    JsonObject query = new JsonObject();
+    JsonObject tagsJson = new JsonObject();
+    for (Map.Entry<String, String> entry : tags.entrySet()) {
+      tagsJson.addProperty(entry.getKey(), entry.getValue());
+    }
+    query.add("tags", tagsJson);
+    JsonArray metricsArray = new JsonArray();
+    for (String metric : metrics) {
+      metricsArray.add(new JsonPrimitive(metric));
+    }
+    query.add("metrics", metricsArray);
+
+    JsonArray groupByArray = new JsonArray();
+    for (String group : groupBy) {
+      groupByArray.add(new JsonPrimitive(group));
+    }
+    query.add("groupBy", groupByArray);
+
+    JsonObject timerangeJson = new JsonObject();
+    for (Map.Entry<String, String> entry : timeRange.entrySet()) {
+      timerangeJson.addProperty(entry.getKey(), entry.getValue());
+    }
+    query.add("timeRange", timerangeJson);
+    return query;
+  }
+
+  class QueryResult {
+    Map<String, String> grouping;
+    String metricName;
+    long numPoints;
+    long totalSum;
+
+    public QueryResult(Map<String, String> grouping, String metricName, long numPoints, long totalSum) {
+      this.grouping = grouping;
+      this.metricName = metricName;
+      this.numPoints = numPoints;
+      this.totalSum = totalSum;
+    }
+
+    public long getTotalSum() {
+      return totalSum;
+    }
+
+    public long getNumPoints() {
+      return numPoints;
+    }
+
+    public String getMetricName() {
+      return metricName;
+    }
+
+    public Map<String, String> getGrouping() {
+      return grouping;
+    }
+  }
+
+
+  private  void batchTest(JsonObject batchRequests,
```

Comment:
will it help to have a method that takes in list of queries and list of expected results and does single query rpc - in url and body, multi query rpc in body? You'd simplify code, potentially a lot